compiler/src/codegen/compcore.re

```diff
@@ -1124,56 +1136,136 @@ let compile_array_op = (wasm_mod, env, arr_imm, op) => {
+            wasm_mod,
+            gensym_label("IndexNotSimpleInteger"),
+            [
+              set_swap(1, load(~offset=4, wasm_mod, get_swap(1))),
+              Expression.Drop.make(
+                wasm_mod,
+                Expression.If.make(
+                  wasm_mod,
+                  Expression.Binary.make(
+                    wasm_mod,
+                    Op.or_int32,
+                    Expression.Binary.make(
+                      wasm_mod,
+                      Op.or_int32,
+                      Expression.Binary.make(
+                        wasm_mod,
+                        Op.eq_int32,
+                        get_swap(1),
+                        Expression.Const.make(
+                          wasm_mod,
+                          const_int32(
+                            tag_val_of_boxed_number_tag_type(BoxedInt32),
+                          ),
+                        ),
+                      ),
+                      Expression.Binary.make(
+                        wasm_mod,
+                        Op.eq_int32,
+                        get_swap(1),
+                        Expression.Const.make(
+                          wasm_mod,
+                          const_int32(
+                            tag_val_of_boxed_number_tag_type(BoxedInt64),
+                          ),
+                        ),
+                      ),
+                    ),
+                    Expression.Binary.make(
+                      wasm_mod,
+                      Op.eq_int32,
+                      get_swap(1),
+                      Expression.Const.make(
+                        wasm_mod,
+                        const_int32(
+                          tag_val_of_boxed_number_tag_type(BoxedBigInt),
+                        ),
+                      ),
+                    ),
+                  ),
+                  call_error_handler(wasm_mod, env, IndexOutOfBounds, []),
+                  call_error_handler(wasm_mod, env, IndexNonInteger, []),
+                ),
+              ),
+            ],
+          ),
+          Expression.Null.make(),
+        ),
+        set_swap(1, untag_number(wasm_mod, get_swap(1))),
+        /* Resolve a negative index */
+        Expression.If.make(
           wasm_mod,
-          env,
           Expression.Binary.make(
             wasm_mod,
-            Op.gt_s_int32,
+            Op.lt_s_int32,
+            get_swap(1),
+            Expression.Const.make(wasm_mod, const_int32(0)),
+          ),
+          set_swap(
+            1,
             Expression.Binary.make(
               wasm_mod,
-              Op.mul_int32,
-              load(~offset=4, wasm_mod, get_arr()),
-              Expression.Const.make(wasm_mod, const_int32(-1)),
+              Op.add_int32,
+              get_swap(1),
+              load(~offset=4, wasm_mod, get_swap(2)),
             ),
-            get_idx(),
           ),
-          IndexOutOfBounds,
-          [],
+          Expression.Null.make(),
```

Comment:
Why do both of these end with a `Null.make()`? Is it to typecheck in wasm or something?

Reply:
Binaryen does one-sided ifs by passing null as the false branch. We could probably change the API to accept an option, though. But I don't really mind this API.